### DISCUSSION TITLE
Compiler warning for parameterless constructors?

### DISCUSSION THREAD

**Discussion by acaly** [dotnet/csharplang#5291] [Q&A] [answered]
Is there plan to add warning to initialize structs without using the parameterless constructor? For example, I think it would be useful to have:

```c#
struct A
{
    public int a = 1;
}

A a1 = default; //warning
A a2 = default!; //OK

class B
{
    private A a1; //warning
    private A a2 = new(); //OK
}

```

I am not sure whether assigning `default` producing error is good, but I do think it's error prone.

**Comment by huoyaoyuan** [comment, marked as answer]
Your example looks more related to non-defaultable structs. I think the scenario should be covered then.

**Reply by acaly** [reply]
Thanks! Is there a proposal for the non-defaultable structs?

**Reply by huoyaoyuan** [reply]
https://github.com/dotnet/csharplang/issues/146

**Reply by RikkiGibson** [reply]
I keep seeing mention of this come up so I decided to post what I've come up with so far in #5337.
